astropy/table/table.py

```diff
@@ -1011,11 +1011,13 @@ def add_index(self, colnames, engine=None, unique=False):
                 raise ValueError('Cannot create an index on column "{}", of '
                                  'type "{}"'.format(col.info.name, type(col)))
 
+        is_primary = not self.indices
         index = Index(columns, engine=engine, unique=unique)
-        if not self.indices:
+        sliced_index = SlicedIndex(index, slice(0, 0, None), original=True)
+        if is_primary:
             self.primary_key = colnames
         for col in columns:
-            col.info.indices.append(index)
+            col.info.indices.append(sliced_index)
```

Comment:
This part is hard to judge without actually understanding the code...

Reply:
The old code for `Index.__new__()` was (effectively):
```
    def __new__(cls, *args, **kwargs):
        self = super().__new__(cls)
        self.__init__(*args, **kwargs)
        return SlicedIndex(self, slice(0, 0, None), original=True)
```
I won't say that I understand all the code, but I do believe that the new code here is equivalent to what the old code was doing.  🤞 that the testing is good enough.